### PR TITLE
Feature cart coupon

### PR DESCRIPTION
**Pull request by rider3458**
coupon "COSC2430-HD" and "COSC2430-DI" can be added to get 20% and 10% discount respectively
applying the wrong coupon code returns an invalid msg
reload page if remove coupon
restyle cart a bit to fit coupon input

### CODE REVIEW

**Review by miketvo — CHANGES_REQUESTED** [review]
- Please use `#ccc5b9` for the "Apply" button background (text color `#403d39`). It is too similar to the "Continue shopping button" at the moment.
- No hover feedback for the "Apply" button. **DO NOT MAKE ME REMIND YOU THIS AGAIN** For the :hover background color, use `#403d39` (text color: `#ccc5b9`).
- USD currency always have two-digit floating point precision (e.g. $12.20, not ~~$12.2~~). Do your research thoroughly, please.

File: public/css/cart.css [hunk 1]
```diff
@@ -239,9 +239,69 @@ input[type=number] {
...
+    border: 1px solid;
+    text-align: center;
+}
+
+.apply-coupon-button,.remove-coupon-button.active {
+    background: #ccc5b9;
+    color: #403d39;
+    display: inline;
+    height: fit-content;
+
```
Comment (doanyennhi) [inline public/css/cart.css:272]: Remove extra line here

**Review by doanyennhi — CHANGES_REQUESTED** [review]

File: public/mall/cart/index.html [hunk 2]
```diff
@@ -110,7 +110,18 @@ <h1 class="content-title">CART</h1>
                 </li>
                 <li class="cart-product-delivery">
                   <span class="cart-product-total-tag">Delivery</span>
-                  <span class="cart-product-total-fee">$15.0</span>
+                  <span class="cart-product-total-fee">$15.00</span>
+                </li>
+                <li class="cart-product-coupon">
+                  <span class="cart-product-total-tag">
+                    <p>Coupon</p>
+                    <label for="insert-coupon"></label>
```
Comment (doanyennhi) [inline public/mall/cart/index.html:118]: Should put "Coupon" inside label

**Review by doanyennhi — CHANGES_REQUESTED** [review]

File: public/mall/cart/index.html (hunk 2 above)
Comment (miketvo) [reply public/mall/cart/index.html:118]: E.g.
```html
<label for="insert-coupon">Coupon</label>
```

**Review by tnathu-ai — COMMENTED** [review]
Disable "Order Placement" page to be **unavailable for guests**, 

Change the styling of coupon message, can try mine:
<p class="coupon-msg" **style="margin: -50px 90px;"**>Coupon applied</p>

Add more meaningful message like: **10% discount is applied, 20% discount** is applied instead of “Coupon applied”

Implement function: When the **quantity of a product is updated through the + and - button** below the price (cart page), the total order amount & price needs to be updated automatically.

**Review by doanyennhi — APPROVED** [review]

**Review by miketvo — APPROVED** [review]

**Pull request merged**